cas-compute/src/symbolic/derivative.rs

```diff
@@ -0,0 +1,268 @@
+use rug::Integer;
+
+use crate::primitive::{float, int};
+
+use super::expr::Primary;
+use super::Expr;
+
+// Something that is not trivially zero - intended to clean up ASTs and not mathetmatically
+// rigorous
+fn non_zero(e: &Expr) -> bool {
+    match e {
+        Expr::Primary(Primary::Float(f)) => {
+            !f.is_zero()
+        },
+        Expr::Primary(Primary::Integer(i)) => {
+            !i.is_zero()
+        },
+        Expr::Add(sum) => {
+            sum.iter().any(non_zero)
+        },
+        Expr::Mul(mul) => {
+            mul.iter().all(non_zero)
+        }
+        _ => true
+    }
```

Comment:
```suggestion
    match e {
        Expr::Primary(Primary::Float(f)) => !f.is_zero(),
        Expr::Primary(Primary::Integer(i)) => !i.is_zero(),
        Expr::Add(sum) => sum.iter().any(non_zero),
        Expr::Mul(mul) => mul.iter().all(non_zero),
        _ => true,
    }
```

- Prefer one-line match arms if the arms are "reasonably" short.
- Possibly missing an `Expr::Exp` match arm?